clang/lib/AST/Expr.cpp

```diff
@@ -222,6 +222,49 @@ bool Expr::isFlexibleArrayMemberLike(
                                          IgnoreTemplateOrMacroSubstitution);
 }
 
+namespace {
+
+/// MemberExprVisitor - Find the MemberExpr through all of the casts, array
+/// subscripts, and unary ops. This intentionally avoids all of them because
+/// we're interested only in the MemberExpr to check if it's a flexible array
+/// member.
+class MemberExprVisitor
+    : public ConstStmtVisitor<MemberExprVisitor, const Expr *> {
+public:
+  //===--------------------------------------------------------------------===//
+  //                            Visitor Methods
+  //===--------------------------------------------------------------------===//
+
+  const Expr *Visit(const Expr *E) {
+    return ConstStmtVisitor<MemberExprVisitor, const Expr *>::Visit(E);
+  }
+  const Expr *VisitStmt(const Stmt *S) { return nullptr; }
+
+  const Expr *VisitMemberExpr(const MemberExpr *E) { return E; }
+
+  const Expr *VisitArraySubscriptExpr(const ArraySubscriptExpr *E) {
+    return Visit(E->getBase());
+  }
+  const Expr *VisitCastExpr(const CastExpr *E) {
+    return Visit(E->getSubExpr());
+  }
+  const Expr *VisitParenExpr(const ParenExpr *E) {
+    return Visit(E->getSubExpr());
+  }
+  const Expr *VisitUnaryAddrOf(const UnaryOperator *E) {
+    return Visit(E->getSubExpr());
+  }
+  const Expr *VisitUnaryDeref(const UnaryOperator *E) {
+    return Visit(E->getSubExpr());
+  }
```

Comment:
> Wouldn’t it be enough to just accept a (possibly parenthesised) `MemberExpr`?

I just saw that you also have `&foo.x[0]` as a test case, but given that this is a new builtin that no compiler supports (yet), is it too much to ask for to require people to just write `foo.x` (or maybe e.g. `(foo.x)` because macro shenanigans)? ;Þ

Reply:
I worry about making assumptions of uses within C. :-) I'm sure that 99.999% of the time, either `foo->x` or `&foo->x[0]` will be the argument. But in the marginal case where it's "hidden" by casts (et al) it worries me that the behavior becomes unknowable. Maybe I'm being too paranoid? How do other places in the front-end handle looking down to the underlying object? Is it mainly through the `Ignore.*` methods?

Reply:
I probably would have just checked if it’s either a `MemberExor`—or possibly a `UnaryOperator` w/ `UO_AddrOf` that contains an `ArraySubscriptExpr` whose LHS is a `MemberExpr`, with a `IgnoreParenImpCasts` between checking for each of these (if you really want to support `&foo->x[0]` too, but are there any real-world use cases for that?). 

(I don’t think there are *so* many different patterns here that we’d need a visitor because the way it’s written rn, it can also lead to false positives, e.g. `__builtin_get_counted_by(**foo->x)`, which seems like it should be nonsense to me)

If, as the result of that, you end up w/ a `MemberExpr`, then evaluate the builtin (i.e. return `null` or the count).

If you *don’t* have a `MemberExpr`, i.e. the user entered garbage like `*foo->x` or something else entirely, or if it is a `MemberExpr` that doesn’t point to a FAM, then I would straight-up error.

> it worries me that the behavior becomes unknowable.

Doing that should also take care of this: if you pass in what boils down to a `MemberExpr` to a FAM, you get either `null` or a pointer to the count; if you pass in anything else, you get an error.

Imo that behaviour is reasonable enough. I don’t think ‘ok, here is something that somehow remotely resembles or involves a member access, so please do *something* with it’ is something we should support—especially because this could ‘silently’ fail in that case if the user did something wrong (by returning null). I’d like to minimise the possibility of that happening.

Also, this is a new builtin, so it’s not like we need to be too worried about breaking anyone’s code because no-one is using it because it’s new. That said, if you end up running into a lot of instances of one particular pattern in e.g. the kernel that this doesn’t support, then it’d make perfect sense to add that too—I just feel like a conservative approach to what we accept would make more sense for the time being.

Reply:
> If you *don’t* have a `MemberExpr`,

Or a `CXXDependentScopeMemberExpr` (I think) if the LHS is dependent, but the rest of the AST should still be the same I believe, even in a template.

Reply:
Oh, on that note, I just saw we’re missing tests for templates; considering that this is supposed to be used primarily in C, we could handle that in a follow-up pr, but I think we should probably support that too at some point (because we allow FAMs in C++ as an extension)

Reply:
> Luckily, the __counted_by__ attribute is C-only, so we won't have to deal with any C++ madnesses. :-)

Ah, I missed that, in that case you can ignore my comment about dependent member expressions.

Reply:
Okay, I removed the Visitor and adopted your suggestion.